README.md
# carmine-sentinel

A Clojure library designed to connect redis by [sentinel](redis.io/topics/sentinel), make [carmine](https://github.com/ptaoussanis/carmine) to support [sentinel](redis.io/topics/sentinel)。

## Usage

```clojure
[net.fnil/carmine-sentinel "1.0.0"]
```

**Carmine-sentinel require carmine version must be `>= 2.15.0`right now.**

First, require carmine and carmine-sentinel:

```clojure
(ns my-app
  (:require [taoensso.carmine :as car]
            [carmine-sentinel.core :as cs :refer [set-sentinel-groups!]]))
```

The only difference compares with carmine is that we will use `carmine-sentinel.core/wcar` to replace `taoensso.carmine/wcar` and add a new function `set-sentinel-groups!`.

Second, configure sentinel groups:

```clojure
(set-sentinel-groups!
  {:group1
   {:specs [{:host "127.0.0.1" :port 5000} {:host "127.0.0.1" :port 5001} {:host "127.0.0.1" :port 5002}]
    :pool  {<opts>} }})
```

There is only one group named `:group1` above, and it has three sentinel instances (port from 5000 to 5002 at 127.0.0.1). Optional, you can set the pool option values and add more sentinel groups.

You can use `add-sentinel-groups!` and `remove-sentinel-group!` to manage the configuration all the time.

Next, we can define the `wcar*`:

```clojure
(def server1-conn {:pool {<opts>} :spec {} :sentinel-group :group1 :master-name "mymaster"})
(defmacro wcar* [& body] `(cs/wcar server1-conn ~@body))
```

The spec in `server1-conn` can be left empty or contain general configurations, such as password or ssl function, and there are two new options in server1-conn:

* `:sentinel-group` Which sentinel instances group to resolve master addr.Here is `:group1`.
* `:master-name` Master name configured in that sentinel group.Here is `mymaster`.

The `spec` in server1-conn will be merged to resolved master spec at runtime.
So you can set `:password`,`:timeout-ms` etc. other options in it.

Also, you can define many `wcar*`-like macros to use other sentinel group and master name.

At last, you can use `wcar*` as the same in carmine.

```clojure
(wcar* (car/set "key" 1))
(wcar* (car/get "key"))
```

If you want to bypass sentinel and connect to redis server directly such as doing testing on your local machine, you can ignore `sentinel-group` and `master-name`, just provide redis server connection spec you want connect to directly like this:

```clojure
(def server1-conn {:pool {<opts>} :spec {:host "127.0.0.1" :port 6379}})
(defmacro wcar* [& body] `(cs/wcar server1-conn ~@body))
```

### Authentication

Due to a bug fix in version `1.0.0` authentication requires slight modifications to the settings.
Notice both the server connection and sentinel group require passing the authentication token:

```clojure
(let [token "foobar"
          host "127.0.0.1"]

      (def server1-conn
        {:pool {}
         :spec {:password token}
         :sentinel-group :group1
         :master-name "mymaster"})

      (set-sentinel-groups!
       {:group1
        {:specs [{:host host :port 5000 :password token}
                 {:host host :port 5001 :password token}
                 {:host host :port 5002 :password token}]}}))
```

`wcar*` is defined normally

## Pub/Sub

Please use `carmine-sentinel.core/with-new-pubsub-listener` to replace `taoensso.carmine/with-new-pubsub-listener` and provide `master-name`, `sentinel-group` to take advantage of sentinel cluster like this:

```clojure
(def server1-conn {:sentinel-group :group1 :master-name "mymaster"})

;;Pub/Sub
(def listener
  (with-new-pubsub-listener server1-conn
    {"foobar" (fn f1 [msg] (println "Channel match: " msg))
     "foo*"   (fn f2 [msg] (println "Pattern match: " msg))}
   (car/subscribe  "foobar" "foobaz")
   (car/psubscribe "foo*")))
```

`carmine-sentinel.core/with-new-pubsub-listener` also support bypass sentinel and connect to redis server directly. You just need to provide the redis server spec you want connect to while ignore `sentinel-group` and `master-name`:

```clojure
(def server1-conn {:spec {:host "127.0.0.1" :port 6379}})

;;Pub/Sub
(def listener
  (with-new-pubsub-listener server1-conn
    {"foobar" (fn f1 [msg] (println "Channel match: " msg))
     "foo*"   (fn f2 [msg] (println "Pattern match: " msg))}
   (car/subscribe  "foobar" "foobaz")
   (car/psubscribe "foo*")))
```

## MessageQueue and Lock

You have to invoke `update-conn-spec` before using other APIs in carmine:

```clojure
(def server1-conn {:pool {<opts>} :spec {} :sentinel-group :group1 :master-name "mymaster"})


;;Message queue
(def my-worker
  (car-mq/worker (cs/update-conn-spec server1-conn) "my-queue"
   {:handler (fn [{:keys [message attempt]}]
               (println "Received" message)
               {:status :success})}))


;;;Lock
(locks/with-lock (cs/update-conn-spec server1-conn) "my-lock"
  1000 ; Time to hold lock
  500  ; Time to wait (block) for lock acquisition
  (println "This was printed under lock!"))
```

## Reading From Slaves

If you want to read data from slave, you can set `prefer-slave?` to be true:

```clojure
(def slave-conn {:pool {<opts>} :spec {}
                 :sentinel-group :group1 :master-name "mymaster"
                 :prefer-slave? true})

(defmacro wcars* [& body] `(cs/wcar slave-conn ~@body))

(wcars* (car/set "key" 1)) ;; ExceptionInfo READONLY You can't write against a read only slave
```

If you have many slaves for one master, the default balancer is `first` function, but you can custom it by `slaves-balancer`,
for example, using random strategy:

```clojure
(def slave-conn {:pool {<opts>} :spec {}
                 :sentinel-group :group1
                 :master-name "mymaster"
                 :prefer-slave? true
                 :slaves-balancer rand-nth)
```

## Listen on carmine-sentinel events

You can register a listener to listen carmine-sentinel events such as `error`, `get-master-addr-by-name`
and `+switch-master` etc. :

```clojure
(cs/register-listener! (fn [e] (println "Event " e " happens")))
```

## Failover

At startup, carmine-sentinel will connect the first sentinel instance to resolve the master address, if if fails, carmine-sentinel will try the next sentinel until find a resolved master address or throw an exception.The resolved addr will be cached in memory.

And Carmine-sentinel subcribes `+switch-master` channel in sentinel.When the master redis instance is down, sentinel will publish a `+switch-master` message, while carmine-sentinel receives this message, it will clean the last cached result and try to connect the new redis master at once.

At last, carmine-sentinel will refresh the sentinel instance list by the response from command `SENTINEL sentinels [name]`.

## API docs

* [Carmine-sentinel APIs](http://fnil.net/docs/carmine_sentinel/)

## Testing

Running the Makefile tests requires having `make`, `lein` and a Linux or MacOS machine.
The test scenario in the is comprised of three sentinels and one master.

To run the tests simply run in shell:
```shell
make test```

## License

Copyright © 2016 killme2008

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log

## [0.3.0] - 2019-02-27
### Compatibility Breaking Change
- Change defined sentinel commands format to be compatible with `carmine` version post `2.14`: will only work for `2.15`+ !

### Fixed
- carmine server connection settings are no longer lost: password connection works!

### Changed
- Expanded hard to read use of macros
- Change variable names from acronyms to full words from acronyms
- `try-resolve-master-spec` now accepts `server-conn` argument
- `get-sentinel-redis-spec` now checks arguments using `:pre` not `if ... throw`


test/carmine_sentinel/core_test.clj
(ns carmine-sentinel.core-test
  (:require [clojure.test :refer :all]
            [taoensso.carmine :as car]
            [carmine-sentinel.core :refer :all]))


;;; NOTE:
;;; add the following to `redis.conf`:
;;; requirepass foobar

;;; NOTE:
;;; SENTINEL CONF `PORT` SHOULD BE DIFFERENT FOR EACH SENTINEL
;;; MAKE SURE TO REMOVE `myid` OR HARDCODE DIFFERENT IDS FOR EACH SENTINEL
;;; NOTE:
;;; Use different files for each sentinel

;;; > cat sentinel.conf
;;; port 5000
;;; daemonize no
;;; pidfile "/var/run/redis-sentinel.pid"
;;; logfile ""
;;; dir "/tmp"
;;; sentinel deny-scripts-reconfig yes
;;; sentinel monitor mymaster 127.0.0.1 6379 2
;;; protected-mode no
;;; requirepass "foobar"


(def token "foobar")
(def host "127.0.0.1")
(def conn {:pool {}
           :spec {:password token}
           :sentinel-group :group1
           :master-name "mymaster"})

(set-sentinel-groups!
 {:group1
  {:specs [{:host host :port 5000 :password token}
           {:host host :port 5001 :password token}
           {:host host :port 5002 :password token}]}})


(deftest resolve-master-spec
  (testing "Try to resolve the master's spec using the sentinels' specs"
    (is (=
         (let [server-conn     {:pool {},
                                :spec {:password "foobar"},
                                :sentinel-group :group1,
                                :master-name "mymaster"}
               specs           [{:host "127.0.0.1", :port 5002, :password "foobar"}
                                {:host "127.0.0.1", :port 5001, :password "foobar"}
                                {:host "127.0.0.1", :port 5000, :password "foobar"}]
               sentinel-group :group1
               master-name    "mymaster"]
           (@#'carmine-sentinel.core/try-resolve-master-spec
            server-conn specs sentinel-group master-name))
         [{:password "foobar", :host "127.0.0.1", :port 6379} ()])
        )))

(deftest subscribing-all-sentinels
  (testing ""
    (is (=
         (let [sentinel-group :group1
               master-name "mymaster"
               server-conn conn]
           (@#'carmine-sentinel.core/subscribe-all-sentinels
            sentinel-group
            master-name))
         [{:password "foobar", :port 5002, :host "127.0.0.1"}
          {:password "foobar", :port 5001, :host "127.0.0.1"}
          {:password "foobar", :port 5000, :host "127.0.0.1"}]))))

(deftest asking-sentinel-master
  (testing ""
    (is (= (let [sentinel-group :group1
                 master-name "mymaster"
                 server-conn conn]
             (@#'carmine-sentinel.core/ask-sentinel-master sentinel-group
              master-name
              server-conn))
           {:password "foobar", :host "127.0.0.1", :port 6379}))))

(deftest sentinel-redis-spec
  (testing "Trying to get redis spec by sentinel-group and master name"
    (is (= {:password "foobar", :host "127.0.0.1", :port 6379}
           (let [server-conn conn]
             (get-sentinel-redis-spec (:sentinel-group server-conn)
                                      (:master-name server-conn)
                                      server-conn))))))

(try
  (defmacro test-wcar* [& body] `(wcar conn ~@body))
  (catch Exception e
    (println "WARNING: caught exception while defining wcar*,"
             "can occur when re-running tests in the same repl."
             "Please verify and check if it isn't the cause of your tests' failure:"
             e)))


(deftest ping
  (testing "Checking if ping works."
    (is (= "PONG" (test-wcar* (car/ping))))))



src/carmine_sentinel/core.clj
(ns carmine-sentinel.core
  (:require [taoensso.carmine :as car]
            [taoensso.carmine.commands :as cmds])
  (:import (java.io EOFException)))

;; {Sentinel group -> master-name -> spec}
(defonce ^:private sentinel-resolved-specs (atom nil))
;; {Sentinel group -> specs}
(defonce ^:private sentinel-groups (volatile! nil))
;; Sentinel event listeners
(defonce ^:private sentinel-listeners (atom nil))
;; Carmine-sentinel event listeners
(defonce ^:private event-listeners (volatile! []))
;; Locks for resolving spec
(defonce ^:private locks (atom nil))

(defn- get-lock [sg mn]
  (if-let [lock (get @locks (str sg "/" mn))]
    lock
    (let [lock (Object.)
          curr @locks]
      (if (compare-and-set! locks curr (assoc curr (str sg "/" mn) lock))
        lock
        (recur sg mn)))))

(defmacro sync-on [sg mn & body]
  `(locking (get-lock ~sg ~mn)
     ~@body))

;;define commands for sentinel
(cmds/defcommand "SENTINEL get-master-addr-by-name"
  {:fn-name         "sentinel-get-master-addr-by-name"
   :fn-params-fixed [name]
   :fn-params-more  nil
   :req-args-fixed  ["SENTINEL" "get-master-addr-by-name" name]
   :cluster-key-idx 2
   :fn-docstring    "get master address by master name. complexity O(1)"})

(cmds/defcommand "SENTINEL slaves"
  {:fn-name         "sentinel-slaves"
   :fn-params-fixed [name]
   :fn-params-more  nil
   :req-args-fixed  ["SENTINEL" "slaves" name]
   :cluster-key-idx 2
   :fn-docstring    "get slaves address by master name. complexity O(1)"})

(cmds/defcommand "SENTINEL sentinels"
  {:fn-name         "sentinel-sentinels"
   :fn-params-fixed [name]
   :fn-params-more  nil
   :req-args-fixed  ["SENTINEL" "sentinels" name]
   :cluster-key-idx 2
   :fn-docstring    "get sentinel instances by mater name. complexity O(1)"})

(defn- master-role? [spec]
  (= "master"
     (first (car/wcar {:spec spec}
                      (car/role)))))

(defn- make-sure-master-role
  "Make sure the spec is a master role."
  [spec]
  (when-not (master-role? spec)
    (throw (IllegalStateException.
            (format "Spec %s is not master role." spec)))))

(defn- dissoc-in
  [m [k & ks :as keys]]
  (if ks
    (if-let [nextmap (get m k)]
      (let [newmap (dissoc-in nextmap ks)]
        (if (seq newmap)
          (assoc m k newmap)
          (dissoc m k)))
      m)
    (dissoc m k)))

(defmacro silently [& body]
  `(try ~@body (catch Exception _#)))

(defn notify-event-listeners [event]
  (doseq [listener @event-listeners]
    (silently (listener event))))

(defn- handle-switch-master [sg msg]
  (when (= "message" (first msg))
    (let [[master-name old-ip old-port new-ip new-port]
          (clojure.string/split (-> msg nnext first)  #" ")]
      (when master-name
        ;;remove last resolved spec
        (swap! sentinel-resolved-specs dissoc-in [sg master-name])
        (notify-event-listeners {:event "+switch-master"
                                 :old {:host old-ip
                                       :port (Integer/valueOf ^String old-port)}
                                 :new {:host new-ip
                                       :port (Integer/valueOf ^String new-port)}})))))

(defrecord SentinelListener [internal-pubsub-listener stopped-mark]
  java.io.Closeable
  (close [_]
    (reset! stopped-mark true)
    (some->> @internal-pubsub-listener (car/close-listener))))

(defn- subscribe-switch-master! [sg spec]
  (if-let [^SentinelListener sentinel-listener (get @sentinel-listeners spec)]
    (deref (.internal-pubsub-listener sentinel-listener))
    (do
      (let [stop? (atom false)
            listener (atom nil)]
        (swap! sentinel-listeners assoc spec (SentinelListener. listener stop?))
        (future
          (while (not @stop?)
            (silently
              ;; It's unusual to use timeout in redis pub/sub but due to Carmine does not
              ;; support ping/pong test for a connection waiting for an event publishing
              ;; from redis, we do need this to maintain liveness in case redis server
              ;; crash unintentionally. Ref. https://github.com/antirez/redis/issues/420
              (let [spec-with-timeout (update spec :timeout-ms #(or % 10000))
                    f (->> (car/with-new-pubsub-listener spec-with-timeout
                             {"+switch-master" (partial handle-switch-master sg)}
                             (car/subscribe "+switch-master"))
                           (reset! listener)
                           :future)]
                (when (not @stop?)
                  (deref f))))
            (silently (Thread/sleep 1000)))))
      (recur sg spec))))

(defn- unsubscribe-switch-master! [sentinel-spec]
  (silently
    (when-let [^SentinelListener sentinel-listener (get @sentinel-listeners sentinel-spec)]
      (.close sentinel-listener)
      (swap! sentinel-listeners dissoc sentinel-spec)
      true)))

(defn- pick-specs-from-sentinel-raw-states [raw-states]
  (map
   (fn [{:strs [ip port]}]
     {:host ip, :port (Integer/valueOf port)})
   (map (partial apply hash-map) raw-states)))

(defn- subscribe-all-sentinels [sentinel-group master-name]
  (when-let [old-sentinel-specs (not-empty (get-in @sentinel-groups [sentinel-group :specs]))]
    (let [valid-specs
          (set ;; remove duplicate sentinel spec
           (flatten
            (mapv
             (fn [spec]
               (try
                 (conj
                  (map (fn [new-raw-spec] (merge spec new-raw-spec))
                       (pick-specs-from-sentinel-raw-states
                        (car/wcar {:spec spec} (sentinel-sentinels master-name))))
                  spec)
                 (catch Exception _ [])))
             old-sentinel-specs)))
          invalid-specs (remove valid-specs old-sentinel-specs)
          ;; still keeping the invalid specs but append them to tail then
          ;; convert spec list to vector to take advantage of their order later
          all-specs (vec (concat valid-specs invalid-specs))]

      (doseq [spec valid-specs]
        (subscribe-switch-master! sentinel-group spec))

      (vswap! sentinel-groups assoc-in [sentinel-group :specs] all-specs)

      (not-empty all-specs))))

(defn- try-resolve-master-spec [server-conn specs sentinel-group master-name]
  (let [sentinel-spec (first specs)]
    (try
      (when-let [[master slaves]
                 (car/wcar {:spec sentinel-spec} :as-pipeline
                           (sentinel-get-master-addr-by-name master-name)
                           (sentinel-slaves master-name))]
        (let [master-spec (merge (:spec server-conn)
                                 {:host (first master)
                                  :port (Integer/valueOf ^String (second master))})
              slaves (pick-specs-from-sentinel-raw-states slaves)]
          (make-sure-master-role master-spec)
          (swap! sentinel-resolved-specs assoc-in [sentinel-group master-name]
                 {:master master-spec
                  :slaves slaves})
          (make-sure-master-role master-spec)
          (notify-event-listeners {:event "get-master-addr-by-name"
                                   :sentinel-group sentinel-group
                                   :master-name master-name
                                   :master master
                                   :slaves slaves})
          [master-spec slaves]))
      (catch Exception e
        (swap! sentinel-resolved-specs dissoc-in [sentinel-group master-name])
        (notify-event-listeners
         {:event "error"
          :sentinel-group sentinel-group
          :master-name master-name
          :sentinel-spec sentinel-spec
          :exception e})
        ;;Close the listener
        (unsubscribe-switch-master! sentinel-spec)
        nil))))

(defn- choose-spec [mn master slaves prefer-slave? slaves-balancer]
  (when (= :error master)
    (throw (IllegalStateException.
            (str "Specs not found by master name: " mn))))
  (if (and prefer-slave? (seq slaves))
    (slaves-balancer slaves)
    master))

(defn- ask-sentinel-master [sentinel-group master-name
                            {:keys [prefer-slave? slaves-balancer]
                             :as server-conn}]
  (if-let [all-specs (subscribe-all-sentinels sentinel-group master-name)]
    (loop [specs all-specs
           tried-specs []]
      (if (seq specs)
        (if-let [[master-spec slaves]
                 (try-resolve-master-spec server-conn specs sentinel-group master-name)]
          (do
            ;;Move the sentinel instance to the first position of sentinel list
            ;;to speedup next time resolving.
            (vswap! sentinel-groups assoc-in [sentinel-group :specs]
                    (vec (concat specs tried-specs)))
            (choose-spec master-name master-spec slaves prefer-slave? slaves-balancer))
          ;;Try next sentinel
          (recur (next specs)
                 (conj tried-specs (first specs))))
        ;;Tried all sentinel instancs, we don't get any valid specs
        ;;Set a :error mark for this situation.
        (do
          (swap! sentinel-resolved-specs assoc-in [sentinel-group master-name]
                 {:master :error
                  :slaves :error})
          (notify-event-listeners {:event "get-master-addr-by-name"
                                   :sentinel-group sentinel-group
                                   :master-name master-name
                                   :master :error
                                   :slaves :error})
          (throw (IllegalStateException.
                  (str "Specs not found by master name: " master-name))))))
    (throw (IllegalStateException.
            (str "Missing specs for sentinel group: " sentinel-group)))))

;;APIs
(defn remove-invalid-resolved-master-specs!
  "Iterate all the resolved master specs and remove any invalid
   master spec found by checking role on redis.
   Please call this periodically to keep safe."
  []
  (doseq [[group-id resolved-specs] @sentinel-resolved-specs]
    (doseq [[master-name master-specs] resolved-specs]
      (try
        (when-not (master-role? (:master master-specs))
          (swap! sentinel-resolved-specs dissoc-in [group-id master-name]))
        (catch EOFException _
          (swap! sentinel-resolved-specs dissoc-in [group-id master-name]))))))

(defn register-listener!
  "Register listener for switching master.
  The listener will be called with an event:
    {:event \"+switch-master\"
     :old {:host old-master-ip
           :port old-master-port
     :new {:host new-master-ip
           :port new-master-port}}}
  "
  [listener]
  (vswap! event-listeners conj listener))

(defn unregister-listener!
  "Remove the listener for switching master."
  [listener]
  (vswap! event-listeners remove (partial = listener)))

(defn get-sentinel-redis-spec
  "Get redis spec by sentinel-group and master name.
   If it is not resolved, it will query from sentinel and
   cache the result in memory.
   Recommend to call this function at your app startup  to reduce
   resolving cost."
  [sentinel-group master-name
   {:keys [prefer-slave? slaves-balancer]
    :or {prefer-slave? false
         slaves-balancer first}
    :as server-conn}]
  {:pre [(not (nil? sentinel-group))
         (not (empty? master-name))]}
  (if-let [resolved-spec (get-in @sentinel-resolved-specs [sentinel-group master-name])]
    (if-let [s (choose-spec master-name
                            (:master resolved-spec)
                            (:slaves resolved-spec)
                            prefer-slave?
                            slaves-balancer)]
      s
      (throw (IllegalStateException.
              (str "Spec not found: " sentinel-group "/" master-name ", " server-conn))))
    ;;Synchronized on [sentinel-group master-name] lock
    (sync-on sentinel-group master-name
             ;;Double checking
             (if (nil? (get-in @sentinel-resolved-specs [sentinel-group master-name]))
               (ask-sentinel-master sentinel-group master-name server-conn)
               (get-sentinel-redis-spec sentinel-group master-name server-conn)))))

(defn set-sentinel-groups!
  "Configure sentinel groups, it will replace current conf:
   {:group-name {:specs [{ :host host
                          :port port
                          :password password
                          :timeout-ms timeout-ms },
                         ...other sentinel instances...]
                 :pool {<opts>}}}
  The conf is a map of sentinel group to connection spec."
  [conf]
  (doseq [[_ group-conf] @sentinel-groups]
    (doseq [spec (:specs group-conf)]
      (unsubscribe-switch-master! spec)))
  (vreset! sentinel-groups conf)
  (reset! sentinel-resolved-specs nil))

(defn add-sentinel-groups!
  "Add sentinel groups,it will be merged into current conf:
   {:group-name {:specs  [{ :host host
                          :port port
                          :password password
                          :timeout-ms timeout-ms },
                          ...other sentinel instances...]
                 :pool {<opts>}}}
  The conf is a map of sentinel group to connection spec."
  [conf]
  (vswap! sentinel-groups merge conf))

(defn remove-sentinel-group!
  "Remove a sentinel group configuration by name."
  [group-name]
  (doseq [sentinel-spec (get-in @sentinel-groups [group-name :specs])]
    (unsubscribe-switch-master! sentinel-spec))
  (vswap! sentinel-groups dissoc group-name)
  (swap! sentinel-resolved-specs dissoc group-name))

(defn remove-last-resolved-spec!
  "Remove last resolved master spec by sentinel group and master name."
  [sg master-name]
  (swap! sentinel-resolved-specs dissoc-in [sg master-name]))

(defn update-conn-spec
  "Cast a carmine-sentinel conn to carmine raw conn spec.
   It will resolve master from sentinel first time,then cache the result in
   memory for reusing."
  [server-conn]
  (if (and (:sentinel-group server-conn) (:master-name server-conn))
    (update server-conn
            :spec
            merge
            (get-sentinel-redis-spec (:sentinel-group server-conn)
                                     (:master-name server-conn)
                                     server-conn))
    server-conn))

(defmacro wcar
  "It's the same as taoensso.carmine/wcar, but supports
      :master-name \"mymaster\"
      :sentinel-group :default
   in conn for redis sentinel cluster.
  "
  {:arglists '([conn :as-pipeline & body] [conn & body])}
  [conn & sigs]
  `(car/wcar
    (update-conn-spec ~conn)
    ~@sigs))

(defmacro with-new-pubsub-listener
  "It's the same as taoensso.carmine/with-new-pubsub-listener,
   but supports
    :master-name \"mymaster\"
    :sentinel-group :default
   in conn for redis sentinel cluster.

   Please note that you can only pass connection spec like
   hostname and port to taoensso.carmine/with-new-pubsub-listener
   like:

   (taoensso.carmine/with-new-pubsub-listener
     {:host \"127.0.0.1\" :port 6379}
     {... channel and handler stuff ... }
     ... publish and subscribe stuff ... )

   but for with-new-pubsub-listener in carmine-sentinel, you need
   to wrap connection spec with another layer along with master-name
   and sentinel-group to take advantage of sentinel cluster like:

   (carmine-sentinel/with-new-pubsub-listener
     {:spec {:host \"127.0.0.1\" :port 6379}
      :master-name \"mymaster\"
      :sentinel-group :default}
     {... channel and handler stuff ... }
     ... publish and subscribe stuff ... )
  "
  [conn-spec & others]
  `(car/with-new-pubsub-listener
     (:spec (update-conn-spec ~conn-spec))
     ~@others))

(defn sentinel-group-status
  "Get the status of all the registered sentinel groups and resolved redis cluster specs.

   For example, firstly we set sentinel groups:

   (set-sentinel-groups!
     {:group1 {:specs [{:host \"127.0.0.1\" :port 5000}
                       {:host \"127.0.0.1\" :port 5001}
                       {:host \"127.0.0.1\" :port 5002}]}})

   Then do something to trigger the resolving of the redis cluster specs.

   (let [server1-conn {:pool {} :spec {} :sentinel-group :group1 :master-name \"mymaster\"}]
     (wcar server1-conn
       (car/set \"a\" 100)))

   At last we execute (sentinel-group-status), then got things like:

   {:group1 {:redis-clusters [{:master-name \"mymaster\",
                               :master-spec {:host \"127.0.0.1\", :port 6379},
                               :slave-specs ({:host \"127.0.0.1\", :port 6380})}],
             :sentinels [{:host \"127.0.0.1\", :port 5000, :with-active-sentinel-listener? true}
                         {:host \"127.0.0.1\", :port 5001, :with-active-sentinel-listener? true}
                         {:host \"127.0.0.1\", :port 5002, :with-active-sentinel-listener? true}]}}"
  []
  (reduce (fn [cur [group-name sentinel-specs]]
            (assoc cur
              group-name
              {:redis-clusters (->> (get @sentinel-resolved-specs group-name)
                                    (mapv (fn [[master-name specs]]
                                            {:master-name master-name
                                             :master-spec (:master specs)
                                             :slave-specs (:slaves specs)})))
               :sentinels      (->> (:specs sentinel-specs)
                                    (map #(let [^SentinelListener listener (get @sentinel-listeners %)]
                                            (assoc %
                                              :with-active-sentinel-listener?
                                              (and (some? listener)
                                                   (not @(.stopped-mark listener)))))))}))
          {}
          @sentinel-groups))

(comment
  (set-sentinel-groups!
   {:group1
    {:specs [{:host "127.0.0.1" :port 5000} {:host "127.0.0.1" :port 5001} {:host "127.0.0.1" :port 5002}]}})
  (let [server1-conn {:pool {} :spec {} :sentinel-group :group1 :master-name "mymaster"}]
    (println
     (wcar server1-conn
           (car/set "a" 100)
           (car/get "a")))))

(comment

  (do ;; reset environment
    (reset! sentinel-resolved-specs nil)
    (vreset! sentinel-groups nil)
    (reset! sentinel-listeners nil)
    (vreset! event-listeners [])
    (reset! locks nil)

    (let [token "foobar"
          host "127.0.0.1"]

      (def server1-conn
        {:pool {}
         :spec {:password token}
         :sentinel-group :group1
         :master-name "mymaster"})

      (set-sentinel-groups!
       {:group1
        {:specs [{:host host :port 5000 :password token}
                 {:host host :port 5001 :password token}
                 {:host host :port 5002 :password token}]}})))

  (defmacro wcar* [& body] `(wcar server1-conn ~@body))

  (wcar* (car/ping))

  (wcar* (car/set "key" 1))

  (wcar* (car/get "key"))

  )
